email/blocker.go

```diff
@@ -85,83 +83,92 @@ func (b *Blocker) threadedBlockMessages() {
 				return
-			case <-ticker.C:
 			}
-		}
-		first = false
 
-		logger.Debugln("Triggered")
+			// log unblocked messages count
+			numUnblocked := len(toBlock)
+			if numUnblocked == 0 {
+				logger.Debugf("Found %v unblocked messages", numUnblocked)
+				return
+			}
+
+			logger.Infof("Found %v unblocked messages", numUnblocked)
+
+			// loop all emails and block the skylinks they contain
+			for _, email := range toBlock {
+				err := b.blockEmail(email)
+				if err != nil {
+					logger.Errorf("Failed to parse email %v, error %v", email.UID, err)
+				}
+			}
+		}()
 
-		// fetch all unblocked emails
-		toBlock, err := abuseDB.FindUnblocked()
-		if err != nil {
-			logger.Errorf("Failed fetching unparsed emails, error %v", err)
+		select {
+		case <-b.staticContext.Done():
+			logger.Debugln("Blocker context done")
 			return
+		case <-ticker.C:
 		}
+	}
+}
 
-		// log unblocked messages count
-		numUnblocked := len(toBlock)
-		if numUnblocked == 0 {
-			logger.Debugf("Found %v unblocked messages\n", numUnblocked)
-			continue
-		}
-		logger.Infof("Found %v unblocked messages\n", numUnblocked)
+// blockEmail will block the skylinks that are contained in the parse result of
+// the given email.
+func (b *Blocker) blockEmail(email database.AbuseEmail) (err error) {
+	// convenience variables
+	abuseDB := b.staticDatabase
 
-		// loop all emails and parse them
-		for _, email := range toBlock {
-			err = func() (err error) {
-				lock := abuseDB.NewLock(email.UID)
-				err = lock.Lock()
-				if err != nil {
-					return errors.AddContext(err, "could not acquire lock")
-				}
-				defer func() {
-					unlockErr := lock.Unlock()
-					if unlockErr != nil {
-						err = errors.Compose(err, errors.AddContext(unlockErr, "could not release lock"))
-						return
-					}
-				}()
-
-				// parse the email
-				result := b.blockReport(email.ParseResult)
-
-				// sanity check we have a result for every skylink
-				if len(result) != len(email.ParseResult.Skylinks) {
-					return errors.New("block result not defined for every skylink")
-				}
+	// acquire the lock
+	lock := abuseDB.NewLock(email.UID)
+	err = lock.Lock()
+	if err != nil {
+		return errors.AddContext(err, "could not acquire lock")
+	}
 
-				// update the email
-				err = abuseDB.UpdateNoLock(email,
-					bson.D{
-						{"$set", bson.D{
-							{"blocked", true},
-							{"block_result", result},
-						}},
-					},
-				)
-				if err != nil {
-					return errors.AddContext(err, "could not update email")
-				}
-				return nil
-			}()
-			if err != nil {
-				logger.Errorf("Failed to parse email %v, error %v", email.UID, err)
-			}
+	// defer the release
+	defer func() {
+		unlockErr := lock.Unlock()
+		if unlockErr != nil {
+			err = errors.Compose(err, errors.AddContext(unlockErr, "could not release lock"))
+			return
 		}
+	}()
+
+	// block the skylinks from the parse result
+	result := b.blockReport(email.ParseResult)
+
+	// sanity check we have a result for every skylink
```

Comment:
Shouldn't this be part of `blockReport`? Or are there situations where it's fine for that method to generate a report with a different number of results than the number of skylinks?

Reply:
Updated